File: android/app/src/main/java/com/lampstandhq/introibo/data/links/LinkMarkup.kt
package com.lampstandhq.introibo.data.links

import com.lampstandhq.introibo.data.model.strippingEm
import com.lampstandhq.introibo.data.search.DeepLinkTarget

// MARK: - Inline-link markup parser
//
// Scans a body string for `<link target="...">...</link>` boundaries and
// produces a list of TextRun values. NO regex — pure string scanning.
//
// Mirror of:
//   Introibo/Data/Links/LinkMarkup.swift

/** A run of either plain text or a link with display text and target. */
sealed class TextRun {
    data class Text(val text: String) : TextRun()
    data class Link(val text: String, val target: DeepLinkTarget) : TextRun()
}

object LinkMarkup {

    /**
     * Parse inline `<link target="...">...</link>` markup into runs.
     *
     * - Text outside links is passed through [strippingEm].
     * - A body with NO `<link>` tags produces exactly one [TextRun.Text] run
     *   equal to `body.strippingEm`.
     * - If a link's target fails to parse, the inner text becomes a
     *   plain [TextRun.Text] run (graceful degradation).
     */
    fun runs(body: String): List<TextRun> {
        val result = mutableListOf<TextRun>()
        var cursor = 0

        while (cursor < body.length) {
            // Find next "<link " tag
            val openStart = body.indexOf("<link ", cursor)
            if (openStart < 0) break

            // Emit text before the tag
            if (cursor < openStart) {
                val before = body.substring(cursor, openStart).strippingEm
                if (before.isNotEmpty()) {
                    result.add(TextRun.Text(before))
                }
            }

            // Extract target="..." — find 'target="' after "<link "
            val targetAttrStart = body.indexOf("target=\"", openStart + 6)
            if (targetAttrStart < 0) break
            val targetValueStart = targetAttrStart + 8 // length of 'target="'
            val targetAttrEnd = body.indexOf('"', targetValueStart)
            if (targetAttrEnd < 0) break
            val targetStr = body.substring(targetValueStart, targetAttrEnd)

            // Find the closing '>' of the opening tag
            val tagClose = body.indexOf('>', targetAttrEnd)
            if (tagClose < 0) break
            val innerStart = tagClose + 1

            // Find </link>
            val closeTagStart = body.indexOf("</link>", innerStart)
            if (closeTagStart < 0) break
            val innerText = body.substring(innerStart, closeTagStart)

            // Parse the target
            val target = LinkTarget.parse(targetStr)
            if (target != null) {
                result.add(TextRun.Link(text = innerText, target = target))
            } else {
                // Graceful degradation: emit inner text as plain text
                if (innerText.isNotEmpty()) {
                    result.add(TextRun.Text(innerText))
                }
            }

            cursor = closeTagStart + 7 // length of "</link>"
        }

        // Emit any remaining text after the last link
        if (cursor < body.length) {
            val tail = body.substring(cursor).strippingEm
            if (tail.isNotEmpty()) {
                result.add(TextRun.Text(tail))
            }
        }

        return result
    }
}

File: Introibo/Data/Links/LinkMarkup.swift
import Foundation

// MARK: - Inline-link markup parser
//
// Scans a body string for `<link target="...">...</link>` boundaries and
// produces an array of TextRun values. NO regex — pure string scanning.
//
// Mirror of:
//   android/app/src/main/java/com/lampstandhq/introibo/data/links/LinkMarkup.kt

/// A run of either plain text or a link with display text and target.
enum TextRun: Hashable {
    case text(String)
    case link(text: String, target: DeepLinkTarget)
}

enum LinkMarkup {

    /// Parse inline `<link target="...">...</link>` markup into runs.
    ///
    /// - Text outside links is passed through `strippingEm`.
    /// - A body with NO `<link>` tags produces exactly one `.text` run
    ///   equal to `body.strippingEm`.
    /// - If a link's target fails to parse, the inner text becomes a
    ///   plain `.text` run (graceful degradation).
    static func runs(_ body: String) -> [TextRun] {
        var result: [TextRun] = []
        var cursor = body.startIndex

        while cursor < body.endIndex {
            // Find next "<link " tag
            guard let openStart = body.range(of: "<link ", range: cursor..<body.endIndex) else {
                break
            }

            // Emit text before the tag
            if cursor < openStart.lowerBound {
                let before = String(body[cursor..<openStart.lowerBound]).strippingEm
                if !before.isEmpty {
                    result.append(.text(before))
                }
            }

            // Extract target="..." — find 'target="' after "<link "
            guard let targetAttrStart = body.range(of: "target=\"", range: openStart.upperBound..<body.endIndex) else {
                // Malformed: no target attribute. Treat rest as text.
                break
            }
            guard let targetAttrEnd = body.firstIndex(of: "\"", after: targetAttrStart.upperBound) else {
                break
            }
            let targetStr = String(body[targetAttrStart.upperBound..<targetAttrEnd])

            // Find the closing '>' of the opening tag
            guard let tagClose = body.firstIndex(of: ">", after: targetAttrEnd) else {
                break
            }
            let innerStart = body.index(after: tagClose)

            // Find </link>
            guard let closeTag = body.range(of: "</link>", range: innerStart..<body.endIndex) else {
                break
            }
            let innerText = String(body[innerStart..<closeTag.lowerBound])

            // Parse the target
            if let target = LinkTarget.parse(targetStr) {
                result.append(.link(text: innerText, target: target))
            } else {
                // Graceful degradation: emit inner text as plain text
                if !innerText.isEmpty {
                    result.append(.text(innerText))
                }
            }

            cursor = closeTag.upperBound
        }

        // Emit any remaining text after the last link
        if cursor < body.endIndex {
            let tail = String(body[cursor...]).strippingEm
            if !tail.isEmpty {
                result.append(.text(tail))
            }
        }

        // If no runs were produced (empty body), return empty array
        return result
    }
}

// MARK: - String helper: find character after a given index

private extension String {
    func firstIndex(of ch: Character, after start: String.Index) -> String.Index? {
        guard start < endIndex else { return nil }
        var i = start
        while i < endIndex {
            if self[i] == ch { return i }
            i = index(after: i)
        }
        return nil
    }
}
